### PR TITLE
Add progressive backoff function

### LINKED ISSUE #10958: Add progressive backoff for Ensure* plugin helper methods.
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-15267)

See https://github.com/mattermost/mattermost-server/pull/10542#discussion_r277337785 for original request.

The proposal is to add a progressive backoff, which is immediately, after 128ms, 256ms, up to three times, instead of waiting the current second.

We should make this into a function that can be used in all kinds of Ensure* plugin helper methods.

### PR DESCRIPTION
**Pull request by jesperhansen17**
#### Summary
Adds a `ProgressiveRetry` function to `utils` package. This function takes an `BackoffOperation` which it executes up to 4 times (3 retries) with an increasing wait time (`128, 256, 512 ms`) between each execution. If the `operation` return a nil error the execution stops and no more retries is performed.

#### Ticket Link
Fixes #10958

### CODE REVIEW

**Comment by hanzei** [comment]
@jesperhansen17 `/utils/backoff.go' is missing a license header. Would you mind copy pasting one into this file? This will fix the build.

**Comment by jesperhansen17** [comment]
@hanzei, sure no problem. I will add a header to the file.

File: plugin/helpers_bots.go [hunk 1]
```diff
@@ -18,11 +18,24 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
 	}
 
 	// If we fail for any reason, this could be a race between creation of bot and
-	// retreval from anouther EnsureBot. Just try the basic retrieve existing again.
+	// retreval from another EnsureBot. Just try the basic retrieve existing again.
```
Comment (lieut-data) [inline plugin/helpers_bots.go]: Thanks for fixing the typo! While you're in here, can we also fix:
```suggestion
	// retrieval from another EnsureBot. Just try the basic retrieve existing again.
```

File: plugin/helpers_bots.go [hunk 2]
```diff
@@ -18,11 +18,24 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
...
 	defer func() {
 		if retBotId == "" || retErr != nil {
-			time.Sleep(time.Second)
-			botIdBytes, err := p.API.KVGet(BOT_USER_KEY)
+			var err error
+			var botIdBytes []byte
+
+			// this func is the operation that will be retried if
```
Comment (lieut-data) [inline plugin/helpers_bots.go]: nit: I'd propose removing the documentation of how `utils.ProgressiveRetry` works. I'm fine with it being hard-coded for now, and if future clients want to customize the retry logic, we can add a parameter which will self-document what happens.

Personally, I'd get rid of the comment altogether, and just inline the very callback as:
```go
err = utils.ProgressiveRetry(func() error {
    // ...
})
```

File: plugin/helpers_bots.go [hunk 3]
```diff
@@ -18,11 +18,24 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
...
+			// and the wait it 128, 256 and 512 ms
+			retryOperation := func() error {
+				botIdBytes, err = p.API.KVGet(BOT_USER_KEY)
+				if err != nil {
+					return err
+				}
+				return nil
+			}
+			err = utils.ProgressiveRetry(retryOperation)
+
```
Comment (lieut-data) [inline plugin/helpers_bots.go]: Orthogonal to your changes, but would it make sense to update `retErr` with the current error?
```suggestion
if err != nil {
    retErr = err
    return
}
```

File: utils/backoff.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		if attempt >= maxAttempts {
+			return err
+		}
+
+		// Wait until timer is finished before trying again
+		<-t.C
+	}
+}
+
+// NextRetry calculates the duration until next retry
```
Comment (lieut-data) [inline utils/backoff.go]: This comment accurately describes the code, but also duplicates it without adding much value. I'd suggest writing this for a human to consume, as something like:
```suggestion
// NextRetry computes the time to wait before the next retry, starting at 128ms and doubling thereafter.
```

However, I also think this code is short enough that you can just inline it directly into `ProgressiveRetry`. At a minimum, there's currently no reason to export it at the package level.

File: utils/backoff_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+		i++
+		return operationError
+	}
+
+	err := ProgressiveRetry(retryOperation)
+
+	if err == nil || err != operationError {
+		t.Errorf("Expected %v, Got %v", operationError, err)
+	}
+	if i != 4 {
```
Comment (lieut-data) [inline utils/backoff_test.go]: Rather than testing `NextRetry` separately, could we find a way to verify that `ProgressRetry` actually waits between attempts? For example, instead of just incrementing `i`, perhaps we add the current time to a slice, and then just verify that the time between intervals is strictly increasing?

File: utils/backoff_test.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	operationError := errors.New("Operation Failed")
+	retryOperation := func() error {
+		i++
+		return operationError
+	}
+
+	err := ProgressiveRetry(retryOperation)
+
+	if err == nil || err != operationError {
+		t.Errorf("Expected %v, Got %v", operationError, err)
```
Comment (lieut-data) [inline utils/backoff_test.go]: We've taken to using https://godoc.org/github.com/stretchr/testify/assert instead of just the built in Go testing primitives, as they add significant clarity when a test fails. Can you rewrite using same? (e.g. `assert.EqualError(err, ...)`

File: utils/backoff_test.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	}
+	if i != 4 {
+		t.Errorf("Invalid number of attempts: %d", i)
+	}
+}
+
+func TestProgressiveRetryShouldSuccessAfterTwoAttempts(t *testing.T) {
```
Comment (lieut-data) [inline utils/backoff_test.go]: Perhaps these two tests could be a single, `TestProgressiveRetry` and then have subtests for the two paths?

File: utils/backoff.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+			t.Reset(nextRetry)
+		}
+
+		attempt++
+		if attempt >= maxAttempts {
+			return err
+		}
+
+		// Wait until timer is finished before trying again
+		<-t.C
```
Comment (lieut-data) [inline utils/backoff.go]: What's the advantage to using a timer here instead of just `time.Sleep`?

File: utils/backoff.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+	maxAttempts        = 4
+)
+
+// BackoffOperation is executed by Retry.
+// The BackoffOperation will be retried the provided
+// number attempts if returning an error.
+type BackoffOperation func() error
+
+// ProgressiveRetry executes a BackoffOperation and retries the operation 3 times upon error.
+func ProgressiveRetry(bo BackoffOperation) error {
```
Comment (levb) [inline utils/backoff.go]: nit: 1/5 s/bo/tryf/ (or whatever, to match the ^^ type name?)

File: utils/backoff.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+// See License.txt for license information.
+
+package utils
+
+import (
+	"time"
+)
+
+const (
+	backoffBase uint64 = 128
```
Comment (levb) [inline utils/backoff.go]: I'm a C programmer myself, but 1/5 I'd do this using the go type system.

Also, FWIW 128ms seems like a pretty long starting timeout.

```go
var backoffTimeouts = []time.Duration {50*time.Millisecond, 100*time.Millisecond, 200*time.Millisecond, 200*time.Millisecond, ...]

// ...
if n > length(backoffTimeouts) {
    // ... error out
}
timeout = backoffTimeouts[n]
```

File: utils/backoff.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+	maxAttempts        = 4
+)
+
+// BackoffOperation is executed by Retry.
+// The BackoffOperation will be retried the provided
+// number attempts if returning an error.
+type BackoffOperation func() error
```
Comment (levb) [inline utils/backoff.go]: nit: suggest `BackoffFunc` or `RetryFunc` or perhaps even just `TryFunc`

File: utils/backoff.go (hunk 9 above)
Comment (levb) [inline utils/backoff.go]: 1/5 would suggest adding an `retryErrorf func(error) bool` parameter that would decide if the error is to retried (true if nil), it can also log if needed.

I would also 0/5 consider using a multierror wrapper to return them all.

Both can be a separate PR though.

File: utils/backoff_test.go (hunk 5 above)
Comment (levb) [reply utils/backoff_test.go]: See my comment about just using a table of durations, then no code to test either.

File: utils/backoff_test.go (hunk 6 above)
Comment (levb) [reply utils/backoff_test.go]: There's also a `...require` version that interrupts the flow of execution, useful for checking errors for instance.

OT: anyone knows of a unit test generator that uses testify? I've been toying with unit test generator built into VSCode, it's not bad even if the tests are ++complicated. Unfortunately, it uses straight `testing` and I'd prefer that it pre-generated with testify assert/require already in place.

File: utils/backoff_test.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+	err := ProgressiveRetry(retryOperation)
+	if err != nil {
+		t.Errorf("Expected %v, Got %v", nil, operationError)
+	}
+	if i != 2 {
+		t.Errorf("Invalid number of attempts: %d", i)
+	}
+}
+
+func TestNextRetryFallsWithinRange(t *testing.T) {
```
Comment (levb) [inline utils/backoff_test.go]: ditto - use a table to start with! then nothing to test, really...

File: utils/backoff.go (hunk 11 above)
Comment (lieut-data) [reply utils/backoff.go]: Another thought: eliminate the type definition altogether, and just accept `func() error`? Seems self sufficient.

File: utils/backoff.go (hunk 9 above)
Comment (lieut-data) [reply utils/backoff.go]: I can see the value of a `retryErrorf`, but at that point I'd almost ask if we need this whole helper function at all, and perhaps the calling code should just `for attempt := 1; attempt <= 3; attempt++ { ... }` and do any custom handling it requires there. There isn't a /lot/ of helper code here to begin with.

File: plugin/helpers_bots.go (hunk 3 above)
Comment (jesperhansen17) [reply plugin/helpers_bots.go]: Hmm, aren't we overwriting previous set errors then with a possible error from `KVGet`?

For example the error in the following code: `return "", errors.Wrap(createBotErr, "failed to create bot")` might be overwritten. By doing so we will never be able to handle the `creation` error from the calling function.

Would @lieut-data say this is an correct understanding?

File: utils/backoff.go (hunk 9 above)
Comment (jesperhansen17) [reply utils/backoff.go]: Might be a dumb question, but what do you mean with `1/5` and `0/5`?

File: utils/backoff.go (hunk 11 above)
Comment (jesperhansen17) [reply utils/backoff.go]: I think dropping the type definition is a great suggestion, because I agree on what @lieut-data is saying that `func() error` is sufficient.

File: utils/backoff.go (hunk 10 above)
Comment (jesperhansen17) [reply utils/backoff.go]: Yeah, 128ms is quite long. But the timeout in `EnsureBot` was previous 1s. So if I were to change the starting timeout to 50ms as you suggesting I would say that I should increase the number of iterations so the max timeout that can occur is atleast 1s.

File: utils/backoff.go (hunk 9 above)
Comment (lieut-data) [reply utils/backoff.go]: Heh, secret decoder ring required: https://docs.mattermost.com/process/training.html#id8.

File: plugin/helpers_bots.go (hunk 3 above)
Comment (lieut-data) [reply plugin/helpers_bots.go]: That's fair: if we assume the caller wanted the error that kickstarted the whole chain of failures, then yes, we should probably leave `retErr` alone.

File: utils/backoff.go (hunk 8 above)
Comment (lieut-data) [reply utils/backoff.go]: @jesperhansen17, wanted to ping again re: `time.Sleep`. From what I can see, there's no reason to create a timer for this case. It would make sense if we wanted to wait on multiple channels to support, say, an abort, but a `time.Sleep` seems like it would simplify the code nicely.

File: utils/backoff.go [hunk 13]
```diff
@@ -7,47 +7,32 @@ import (
...
+var backoffTimeouts = []time.Duration{50 * time.Millisecond, 100 * time.Millisecond, 200 * time.Millisecond, 200 * time.Millisecond, 400 * time.Millisecond, 400 * time.Millisecond}
 
 // ProgressiveRetry executes a BackoffOperation and retries the operation 3 times upon error.
-func ProgressiveRetry(bo BackoffOperation) error {
+func ProgressiveRetry(operation func() error) error {
 	var t *time.Timer
-	var attempt uint64
+	var attempts = 0
 
 	for {
```
Comment (lieut-data) [inline utils/backoff.go]: Worth just simplifying to:

```go
for attempts := 0; attempts < len(backoffTimeouts); attempts++ {
```

File: utils/backoff.go [hunk 14]
```diff
@@ -7,47 +7,32 @@ import (
...
 			return nil
 		}
 
-		nextRetry := NextRetry(attempt)
+		attempts++
+		if attempts >= len(backoffTimeouts) {
+			return err
+		}
+
+		nextRetry := backoffTimeouts[attempts]
```
Comment (lieut-data) [inline utils/backoff.go]: The first time we index into `backoffTimeouts`, `attempts == 1`, so we never actually wait 50ms: always 100ms. See above for a simpler suggestion that just iterates `attempts` starting from `0`.

File: utils/backoff.go (hunk 8 above)
Comment (jesperhansen17) [reply utils/backoff.go]: In this implementation I can't say there is any, but I used timer before and that's why I choose to use it again. What I do like with it comes as a channel and therefore I can make use of other go routines and I feel it gives better flexibility in handling parallel go routines.

But yeah, here in this case I can't really say what the main advantage is.

File: go.mod [hunk 15]
```diff
@@ -79,10 +79,10 @@ require (
 	go.uber.org/zap v1.9.1
 	golang.org/x/crypto v0.0.0-20190426145343-a29dc8fdc734
 	golang.org/x/image v0.0.0-20190321063152-3fc05d484e9f
-	golang.org/x/net v0.0.0-20190424112056-4829fb13d2c6
-	golang.org/x/sync v0.0.0-20190423024810-112230192c58 // indirect
+	golang.org/x/net v0.0.0-20190620200207-3b0461eec859
```
Comment (lieut-data) [inline go.mod]: @jesperhansen17, any idea why these changed as part your commits? I checked upstream, but couldn't find anything from `master`. Perhaps just revert `go.mod` and `go.sum` unless there's a reason to effect a change here?

Comment (jesperhansen17) [reply go.mod]: I think I accidentally add a commit on `go.mod` and `go.sum` after I ran some of the makefile commands (which generates new `go.mod` and `go.sum` files for me).

I have reverted the changes.
